flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroReaderWriter.java

```diff
@@ -22,52 +22,76 @@
 import java.io.File;
 import java.io.IOException;
 import java.util.Iterator;
-import org.apache.flink.types.Row;
+import java.util.List;
+import org.apache.flink.table.data.RowData;
 import org.apache.iceberg.Files;
 import org.apache.iceberg.Schema;
 import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.data.DataTest;
+import org.apache.iceberg.data.RandomGenericData;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.data.avro.DataWriter;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
 import org.apache.iceberg.io.CloseableIterable;
 import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.junit.Assert;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.TemporaryFolder;
 
-import static org.apache.iceberg.flink.data.RandomData.COMPLEX_SCHEMA;
+public class TestFlinkAvroReaderWriter extends DataTest {
 
-public class TestFlinkAvroReaderWriter {
-  private static final int NUM_RECORDS = 20_000;
+  private static final int NUM_RECORDS = 100;
 
-  @Rule
-  public TemporaryFolder temp = new TemporaryFolder();
+  @Override
+  protected void writeAndValidate(Schema schema) throws IOException {
+    List<RowData> inputs = generateDataFromAvroFile(schema);
 
-  private void testCorrectness(Schema schema, int numRecords, Iterable<Row> iterable) throws IOException {
     File testFile = temp.newFile();
     Assert.assertTrue("Delete should succeed", testFile.delete());
 
-    try (FileAppender<Row> writer = Avro.write(Files.localOutput(testFile))
+    try (FileAppender<RowData> writer = Avro.write(Files.localOutput(testFile))
         .schema(schema)
-        .createWriterFunc(FlinkAvroWriter::new)
+        .createWriterFunc(ignore -> new FlinkAvroWriter(FlinkSchemaUtil.convert(schema)))
         .build()) {
-      writer.addAll(iterable);
+      writer.addAll(inputs);
     }
 
-    try (CloseableIterable<Row> reader = Avro.read(Files.localInput(testFile))
+    try (CloseableIterable<RowData> reader = Avro.read(Files.localInput(testFile))
         .project(schema)
         .createReaderFunc(FlinkAvroReader::new)
         .build()) {
-      Iterator<Row> expected = iterable.iterator();
-      Iterator<Row> rows = reader.iterator();
-      for (int i = 0; i < numRecords; i += 1) {
+      Iterator<RowData> expected = inputs.iterator();
+      Iterator<RowData> rows = reader.iterator();
+      for (int i = 0; i < NUM_RECORDS; i += 1) {
         Assert.assertTrue("Should have expected number of rows", rows.hasNext());
         Assert.assertEquals(expected.next(), rows.next());
       }
       Assert.assertFalse("Should not have extra rows", rows.hasNext());
     }
   }
 
-  @Test
-  public void testNormalData() throws IOException {
-    testCorrectness(COMPLEX_SCHEMA, NUM_RECORDS, RandomData.generate(COMPLEX_SCHEMA, NUM_RECORDS, 19982));
+  private List<RowData> generateDataFromAvroFile(Schema schema) throws IOException {
```

Comment:
I think it would be better to validate Flink `RowData` against generic `Record`. That's what we do in Spark tests, where we first write using generics (or Avro in older tests) and then validate that the records we read using the Spark object model are equivalent. By doing that, you not only test that `RowData` to disk and back to `RowData` works, but that the records are actually equivalent to another read format.

Reply:
You are right, we should have a asserter for `RowData` and `Record`.